### ISSUE TITLE
move modules to his own repository?

### ISSUE THREAD

**Issue by jknack** [jooby-project/jooby#651]
Due that jooby has so many modules and of course we are going to keep adding... may be a good idea to move modules to his own git repository.

All these modules make hard to quickly check jooby core and/or see a quick project overview due we must scroll a lot.. to see it.

Still there is one main drawbacks: maven and release process.  Existing release process works quite well and it is to produce/maintain releases. If we split modules into his own repository the release process will be affected.

Initial suggestion:

1. move all the modules to his own repository
2. create a modules folder in jooby core (this repository)
3. uses git submodules to load modules inside the modules folder
4. create a couple of scripts to assemble, aggregate and release jooby with all modules.

**Comment by noboomu** [comment]
This would make it much easier for others to contribute as well...

**Comment by BodySplash** [comment]
First of all, congratulations for this excellent project. 
 In my experience, submodules are a real PITA to maintain. In another project, I followed mostly your plan, with the difference that the submodules were checkedout via a script, and ignored in the main repository. 
We even have a little shell tool to ease that kind of setup : https://github.com/arpinum-oss/beastmaster.

**Comment by jknack** [comment]
@BodySplash Thank you, appreciate your feedback. I considered Jooby is a great project to work with but it is hard to reach people and build medium size community.

Think will go with submodules as described here, but can you point me to your example? I mean a direct link to to shell script.

Thanks again and please don't forget to star Jooby.

**Comment by BodySplash** [comment]
All I can find in my github repositories is this set of ruby scripts for a former open source project of mine : https://github.com/MongoLink/mongolink-tools 
Their are so simple I'm a little bit ashamed to show them :) 

Concerning submodules, maybe subtrees could be a better solution then. I believe submodules are great to include stable dependencies, but not that great to work on a multi module project, where you want to feel you're working on a single code base. This [article](https://medium.com/@v/git-subtrees-a-tutorial-6ff568381844#.tqactheu4) is a good introduction to this matter.

**Comment by ph-ms** [comment]
@noboomu  @BodySplash - if you had to limit to a single strong reason for NOT having this monorepo design is it:
1. The clone takes too long
2. The .git folder on my machine is too big
3. It is hard to navigate, once cloned
4. I don't want spend the elapsed time to build everything were I to make a small contrib to a relative 'leaf-node' module
5. As #4 but s/elapsed time/local Maven repo disk space/
6. As #4 but s/elapsed time/my connection to the internet isn't cheap and fast for Maven downloading/
7. Something else?

**Comment by BodySplash** [comment]
Clearly number 4. It's also a matter of clarity : I don't have to look at all this code that is valuable, but useless for my own work.

**Comment by ph-ms** [comment]
Thanks Jean-Baptiste. @noboomu ?

**Comment by nesl247** [comment]
I'd definitely move each module to it's own repository. This does allow another benefit as well. Each module could have it's own releases then. Rather than module-x having a breaking bug or something and waiting for all of jooby to have a release, you could easily release a bugfix.

**Comment by nesl247** [comment]
Not trying to spam the thread, but one more reason is different maintainers would be possible then without them having access to merge in PRs for every other part of the framework.

**Comment by jschneider** [comment]
I personally switched to monorepos for my projects. And I love it.
The release process and dependency management is so much simpler...

Of course I see the other arguments. Don't know what is right for Jooby. But I will definitely not look back to separate repos. There is a reason why large companies (Google, Facebook, ...) use mono repositories.

**Comment by gamaleyvv** [comment]
http://gbayer.com/development/moving-files-from-one-git-repository-to-another-preserving-history/
https://help.github.com/articles/splitting-a-subfolder-out-into-a-new-repository/
Sorry: It's hard, but you can ...

**Comment by paul-hammant** [comment]
Full disclosure (I'm ph-ms too - work account): https://paulhammant.com/2017/04/04/benchmarking-a-maven-monorepo/

There's ways of working with Jooby as we have it now, that mean there's less MB consumed on hard drives and lower elapsed times on build/tests of targetted parts of the Jooby ecosystem. I'm also a veteran of Google's monorepo and documentator of it here - https://trunkbaseddevelopment.com/expanding-contracting-monorepos/ (and more).

TL;DR - Jooby's Monorepo is just fine.

**Comment by jschneider** [comment]
One thing from a user perspective:

Typically a user has several jooby modules in his pom. And I am quite sure they want to have just one version number over all modules.
I have had the other case (with assertj). And it is a real pain.
Updating each module on its own. They have common dependencies with different version numbers. That is really, really hard.

Therefore I think it is necessary to release each and every module every time with the same version number.


If you prefer to release each and every module on its own every time, I am ok with separate repositories (but I'd prefer the mono repo).


TL;DR: The users want *one* version number for all jooby modules.

**Comment by paul-hammant** [comment]
I agree, @jschneider. The only downside is that we get to flood Maven central with things that bump version number but are binary identical (ish) to their predecessor. Of course with Edgar's monorepo design, he can cd into (say) jooby-mongodb and to a point release to solve one bug, which incentivized downstream teams can pull in range for all the Jooby pieces ([ref](http://stackoverflow.com/questions/30571/how-do-i-tell-maven-to-use-the-latest-version-of-a-dependency)),

**Comment by jschneider** [comment]
@paul-hammant Yes, we flood Maven central. 
But I'd prefer that over burdening the user with having to do the work itself.
Hard disks are cheap these days... ;-)
